moveit_core/robot_model/src/robot_model.cpp

```diff
@@ -1176,14 +1176,28 @@ LinkModel* RobotModel::getLinkModel(const std::string& name, bool* has_link)
   return nullptr;
 }
 
-const LinkModel* RobotModel::getRigidlyConnectedParentLinkModel(const LinkModel* link)
+const LinkModel* RobotModel::getRigidlyConnectedParentLinkModel(const LinkModel* link, const JointModelGroup* jmg)
 {
   if (!link)
     return link;
   const moveit::core::LinkModel* parent_link = link->getParentLinkModel();
   const moveit::core::JointModel* joint = link->getParentJointModel();
+  decltype(jmg->getJointModels().cbegin()) begin{}, end{};
+  if (jmg)
+  {
+    begin = jmg->getJointModels().cbegin();
+    end = jmg->getJointModels().cend();
+  }
+
+  auto is_fixed = [begin, end](const moveit::core::JointModel* joint) {
+    if (joint->getType() == moveit::core::JointModel::FIXED)
+      return true;
+    if (begin != end && std::find(begin, end, joint) == end)
+      return true;
+    return false;
+  };
 
-  while (parent_link && joint->getType() == moveit::core::JointModel::FIXED)
+  while (parent_link && is_fixed(joint))
```

Comment:
It took me some while to parse and understand above code, maybe split into two lambdas? It's a shame that `contains` is only C++ 20

Reply:
I agree. Splitting into two lambdas (`is_fixed(joint) && in_jmg(joint)`) is not possible, because that would change the logic. But I renamed the lambda and added some comments to clarify semantics. Hope that helps.

Reply:
it would be `parent_link && (is_fixed(joint) || ! in_jmg(joint))` or `parent_link && (is_fixed(joint) || ! jmg->hasJoint(joint.name)))`

Reply:
I don't consider this complex expression easier to parse :wink: